File: Chapter 1/1.1/exercises/1.7.js
//Program to calculate Square Root
//Find the square root
function sqrt(x) {
  //Start guessing
  sqrtIter(1.0, x);
}

//Improve a guess
function improve(guess, x) {
  return (guess + x / guess) / 2;
}

function goodEnough(guess, x) {
  if (improve(guess, x) - guess === 0) {
    return true;
  }

  return false;
}

function sqrtIter(g, x) {
  console.log(g);
  //Base Case
  if (goodEnough(g, x)) {
    console.log("");
    return g;
  }

  //New Guess
  let newg = improve(g, x);
  //Repeat
  return sqrtIter(newg, x);
}

let r = sqrt(2);

console.log(r);

File: Chapter 1/1.1/fun/1.7.c
#include <stdbool.h>
#include <stdio.h>

//Improve a guess
double improve(double guess, double x) {
    return (guess + x / guess) / 2;
}

bool goodEnough(double guess, double x) {
    double newGuess = improve(guess, x);

    if (newGuess - guess == 0) {
        return true;
    }

    return false;
}

double sqrtIter(double g, double x) {
    printf("%f\n", g);
    //Base Case
    if (goodEnough(g, x)) {
        printf("\n");
        return g;
    }

    //New Guess
    double newg = improve(g, x);
    //Repeat
    sqrtIter(newg, x);
}
double my_sqrt(double x) {
    //Start guessing
    double ans;
    ans = sqrtIter(1, x);
    return ans;
}

double main() {
    double x = 2;
    double r = my_sqrt(x);
    printf("%f\n", r);
    return r;
}